core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala

```diff
@@ -405,4 +405,90 @@ class TSqlExpressionBuilder(functionBuilder: FunctionBuilder)
     val dataType = dataTypeBuilder.build(ctx.dataType())
     ir.Cast(expression, dataType, returnNullOnError = ctx.TRY_CAST() != null)
   }
+
+  override def visitJsonArray(ctx: JsonArrayContext): ir.Expression = {
+    val elements = buildExpressionList(Option(ctx.expressionList()))
+    val absentOnNull = checkAbsentNull(ctx.jsonNullClause())
+    buildJsonArray(elements, absentOnNull)
+  }
+
+  override def visitJsonObject(ctx: JsonObjectContext): ir.Expression = {
+    val namedStruct = buildNamedStruct(Option(ctx.jsonKeyValue().asScala.toList))
+    val absentOnNull = checkAbsentNull(ctx.jsonNullClause())
+    buildJsonObject(namedStruct, absentOnNull)
+  }
+
+  private def checkAbsentNull(ctx: JsonNullClauseContext): Boolean = {
+    ctx match {
+      case null => true
+      case clause if clause.ABSENT() == null => false
+      case _ => true
+    }
+  }
+
+  private def buildNamedStruct(ctx: Option[List[JsonKeyValueContext]]): ir.NamedStruct = {
+    val (keys, values) = ctx
+      .getOrElse(List.empty)
+      .map { keyValueContext =>
+        val expressions = keyValueContext.expression().asScala.toList
+        (expressions.head.accept(this), expressions(1).accept(this))
+      }
+      .unzip
+
+    ir.NamedStruct(keys, values)
+  }
```

Comment:
```suggestion
  override def visitJsonObject(ctx: JsonObjectContext): ir.Expression = {
    val jsonKeyValues = Option(ctx.jsonKeyValue()).map(_.asScala).getOrElse(Nil)
    val namedStruct = buildNamedStruct(jsonKeyValues)
    val absentOnNull = checkAbsentNull(ctx.jsonNullClause())
    buildJsonObject(namedStruct, absentOnNull)
  }

  private def checkAbsentNull(ctx: JsonNullClauseContext): Boolean = {
    ! Option(ctx).exists(_.ABSENT() != null)
  }

  private def buildNamedStruct(ctx: Seq[JsonKeyValueContext]): ir.NamedStruct = {
    val (keys, values) = ctx
      .map { keyValueContext =>
        val expressions = keyValueContext.expression().asScala.toList
        (expressions.head.accept(this), expressions(1).accept(this))
      }
      .unzip

    ir.NamedStruct(keys, values)
  }
```

Reply:
While a good idea, the code:

```scala
  private def checkAbsentNull(ctx: JsonNullClauseContext): Boolean = {
    ctx match {
      case null => true
      case clause if clause.ABSENT() == null => false
      case _ => true
    }
  }
  ```
  
  Is required because:
  
    - If the clause does not exist, the ABSENT ON NULL is assumed - so true (case null prevents null access)
    - If the clause exists and ABSENT == null then the clause was NULL ON NULL - so false
    - If the clause exists and ABSENT != null then the clause was ABSENT ON NULL - so true
    
    Your code:
    
 ```scala
    private def checkAbsentNull(ctx: JsonNullClauseContext): Boolean = {
    ! Option(ctx).exists(_.ABSENT() != null)
  }
```
 
 Does not implement the same logic and it took me a minute to see why. I will add the above as a comment.